vault/src/metrics.rs

```diff
@@ -0,0 +1,527 @@
+            },
+        }
+    }
+
+    pub async fn initialize_values<B: BitcoinCoreApi + Clone + Send + Sync>(
+        parachain_rpc: InterBtcParachain,
+        vault_id: VaultId,
+        vault: &VaultData<B>,
+    ) {
+        let bitcoin_transactions = match vault.btc_rpc.list_transactions(None).await {
+            Ok(x) => x
+                .into_iter()
+                .filter(|x| x.detail.category == GetTransactionResultDetailCategory::Send)
+                .collect(),
+            Err(_) => vec![],
+        };
+
+        // update average fee
+        let (total, count) = bitcoin_transactions
+            .iter()
+            .filter_map(|tx| tx.detail.fee.map(|amount| amount.as_sat().abs() as u64))
+            .fold((0, 0), |(total, count), x| (total + x, count + 1));
+        *vault.metrics.average_btc_fee.data.write().await = AverageTracker { total, count };
+
+        // update fee surplus
+        if let Ok((redeem_requests, replace_requests, refund_requests)) = try_join!(
+            parachain_rpc.get_vault_redeem_requests(vault_id.account_id.clone()),
+            parachain_rpc.get_old_vault_replace_requests(vault_id.account_id.clone()),
+            parachain_rpc.get_vault_refund_requests(vault_id.account_id.clone()),
+        ) {
+            let redeems = redeem_requests
+                .iter()
+                .map(|(id, redeem)| (id.clone(), redeem.transfer_fee_btc));
+            let refunds = refund_requests
+                .iter()
+                .map(|(id, refund)| (id.clone(), refund.transfer_fee_btc));
+            let replaces = replace_requests.iter().map(|(id, _)| (id.clone(), 0));
+            let fee_budgets = redeems.chain(refunds).chain(replaces).collect::<HashMap<_, _>>();
+            let fee_budgets = &fee_budgets;
+
+            let fee_budget_surplus = futures::stream::iter(bitcoin_transactions.iter())
+                .filter_map(|tx| async move {
+                    let transaction = vault
+                        .btc_rpc
+                        .get_transaction(&tx.info.txid, tx.info.blockhash)
+                        .await
+                        .ok()?;
+                    let op_return = transaction.get_op_return()?;
+                    let budget: i64 = fee_budgets.get(&op_return)?.clone().try_into().ok()?;
+                    let surplus = budget.checked_sub(tx.detail.fee?.as_sat().abs());
+                    surplus
+                })
+                .fold(0i64, |acc, x| async move { acc.saturating_add(x) })
+                .await;
+
+            *vault.metrics.fee_budget_surplus.data.write().await = fee_budget_surplus;
+            publish_fee_budget_surplus(&vault).await;
+        }
+
+        publish_utxo_count(&vault).await;
+        publish_average_bitcoin_fee(&vault).await;
+        publish_bitcoin_balance(&vault).await;
+        update_expected_bitcoin_balance(&vault, parachain_rpc.clone()).await;
```

Comment:
Not really worth it - as you noted, two of these are not actually async due to bitcoin rpc being sync, and one of the others is only async because of the rwlock